base/reducedim.jl

```diff
@@ -359,17 +359,21 @@ mapreduce(f, op, A::AbstractArrayOrBroadcasted; dims=:, init=_InitialValue()) =
 mapreduce(f, op, A::AbstractArrayOrBroadcasted...; kw...) =
     reduce(op, map(f, A...); kw...)
 
-_mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, ::Colon) =
-    mapfoldl_impl(f, op, nt, A)
-
-_mapreduce_dim(f, op, ::_InitialValue, A::AbstractArrayOrBroadcasted, ::Colon) =
-    _mapreduce(f, op, IndexStyle(A), A)
-
-_mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, dims) =
-    mapreducedim!(f, op, reducedim_initarray(A, dims, nt), A)
+function _mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, ::Colon)
+    if nt isa _InitialValue
+        _mapreduce(f, op, IndexStyle(A), A)
+    else
+        mapfoldl_impl(f, op, nt, A)
+    end
+end
 
-_mapreduce_dim(f, op, ::_InitialValue, A::AbstractArrayOrBroadcasted, dims) =
-    mapreducedim!(f, op, reducedim_init(f, op, A, dims), A)
+function _mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, dims)
+    if nt isa _InitialValue
+        mapreducedim!(f, op, reducedim_init(f, op, A, dims), A)
+    else
+        mapreducedim!(f, op, reducedim_initarray(A, dims, nt), A)
+    end
+end
```

Comment:
No need to change these?

Reply:
Without this change, I would have to define two separate methods for `_mapreduce_dim` for `FastSubArray`, one with `_InitialValue` and one without. Moreover, the method with `_InitialValue` would have to go to reducedim.jl instead of subarray.jl because `_InitialValue` is not yet defined when subarray.jl is read. I wanted to keep all the `SubArray` code in subarray.jl.

However, with your idea to change `_mapreduce` instead, no change would be necessary in reducedim.jl.